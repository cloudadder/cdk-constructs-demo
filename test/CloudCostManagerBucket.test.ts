import { App } from 'aws-cdk-lib';
import { Annotations, Template } from 'aws-cdk-lib/assertions';
import { ExistingBucketAfterAddingTiering, ExistingBucketWithErrors, TestStackBucket, TestStackBucketWithErrors } from './TestStacks';

describe('Cloud Cost Manager Bucket', () => {
  test('(BucketTestNegitive): confirm cloud cost management reject stack for bucket', () => {
    const app = new App();
    var stack = new TestStackBucketWithErrors(app, 'TestStackBucket');

    Annotations.fromStack(stack).hasError('/TestStackBucket',
      'Buckets require intelligent tiering',
    );
  });

  test('(BucketTestPositve): confirm cloud cost management is implemented for bucket', () => {
    const app = new App();
    const stack = new TestStackBucket(app, 'TestStackBucket');

    Annotations.fromStack(stack).hasInfo('/TestStackBucket',
      'CloudCostManager validation passed',
    );

    Template.fromStack(stack).hasResourceProperties('AWS::S3::Bucket', {
      Tags: [
        {
          Key: 'cloud-cost-manager:check:bucket',
          Value: 'pass',
        },
        {
          Key: 'cloud-cost-manager:customer-name',
          Value: 'acme-co',
        },
        {
          Key: 'cloud-cost-manager:env-name',
          Value: 'staging',
        },
      ],
    });
  });

  test('(ExistingCFTestNegitive): confirm cloud cost management reject stack for bucket from existing CF', () => {
    const app = new App();
    var stack = new ExistingBucketWithErrors(app, 'TestStackBucket');

    Annotations.fromStack(stack).hasError('/TestStackBucket',
      'Buckets require intelligent tiering',
    );
  });

  test('(ExistingCFTestPositve): confirm cloud cost management is implemented for bucket from existing CF', () => {
    const app = new App();
    const stack = new ExistingBucketAfterAddingTiering(app, 'TestStackBucket');

    Annotations.fromStack(stack).hasInfo('/TestStackBucket',
      'CloudCostManager validation passed',
    );

    Template.fromStack(stack).hasResourceProperties('AWS::S3::Bucket', {
      Tags: [
        {
          Key: 'cloud-cost-manager:check:bucket',
          Value: 'pass',
        },
        {
          Key: 'cloud-cost-manager:customer-name',
          Value: 'acme-co',
        },
        {
          Key: 'cloud-cost-manager:env-name',
          Value: 'staging',
        },
      ],
    });
  });

  test('Snapshot Bucket', () => {
    const app = new App();
    const stack = new TestStackBucket(app, 'test');

    expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
  });
});