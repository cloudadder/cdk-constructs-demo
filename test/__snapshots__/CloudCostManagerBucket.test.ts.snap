// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Cost Manager Bucket Snapshot Bucket 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "testbucketE6E05ABE": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "IntelligentTieringConfigurations": Array [
          Object {
            "Id": "test-tier",
            "Status": "Enabled",
            "Tierings": Array [],
          },
        ],
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:check:bucket",
            "Value": "pass",
          },
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "staging",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
